File: AlgorithmsPackJavaSolutions/Longest Increasing Path in a Matrix.java
class Solution {
    private int[] dr = {-1, 0, 1, 0};
    private int[] dc = {0, 1, 0, -1};

    private class Node {
        int r;
        int c;

        Node(int r, int c) {
            this.r = r;
            this.c = c;
        }
    }

    private boolean isValid(int r, int c, int[][] matrix) {
        if (r < 0 || r >= matrix.length)
            return false;
        if (c < 0 || c >= matrix[0].length)
            return false;
        return true;
    }

    private Pair<List<Node>, Integer> biggerNeighboursAndEdges(int r, int c, int[][] matrix) {
        List<Node> edges = new ArrayList<>();
        int incomingEdges = 0;

        for (int d = 0; d < 4; d++) {
            int nr = dr[d] + r;
            int nc = dc[d] + c;
            if (!isValid(nr, nc, matrix))
                continue;

            if (matrix[r][c] < matrix[nr][nc])
                edges.add(new Node(nr, nc));
            else if (matrix[r][c] > matrix[nr][nc])
                incomingEdges++;
        }

        return new Pair<>(edges, incomingEdges);
    }

    public int longestIncreasingPath(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        int[][] incomingEdges = new int[rows][cols];
        Queue<Node> queue = new LinkedList<>();

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                Pair<List<Node>, Integer> result = biggerNeighboursAndEdges(r, c, matrix);
                incomingEdges[r][c] = result.getValue();
                if (incomingEdges[r][c] == 0)
                    queue.add(new Node(r, c));
            }
        }

        int levels = 0;
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            while (levelSize-- > 0) {
                Node node = queue.poll();
                List<Node> biggerNeighbours = biggerNeighboursAndEdges(node.r, node.c, matrix).getKey();

                for (Node neighbour : biggerNeighbours) {
                    if (--incomingEdges[neighbour.r][neighbour.c] == 0)
                        queue.add(neighbour);
                }
            }
            levels++;
        }

        return levels;
    }
}

File: AlgorithmsPackCPP/Longest Increasing Path in a Matrix.cpp
class Solution {
    int dr[4]{-1,0,1,0};
    int dc[4]{0,1,0,-1};
    
    struct node{
        int r;
        int c;
    };
    
    bool isValid(int r, int c, vector<vector<int>>&matrix){
        if(r < 0 || r>=matrix.size())
            return false;
        if(c<0||c>=matrix[0].size())
            return false;
        return true;
    }
public:
    pair<vector<node>,int>biggerNeighboursAndEdges(int r, int c, vector<vector<int>>&matrix){
        vector<node>edges;
        int incomingEdges = 0;
        
        for(int d = 0; d < 4; d++){
            int nr = dr[d]+r;
            int nc = dc[d]+c;
            if(!isValid(nr,nc,matrix))
                continue;

            if(matrix[r][c] < matrix[nr][nc]) 
                edges.push_back({nr,nc});
            else if(matrix[r][c] > matrix[nr][nc]) 
                incomingEdges++;
        }
        return {edges,incomingEdges}; 
    }
    
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int rows = matrix.size();
        int cols = matrix[0].size();
        vector<vector<int>>incomingEdges(rows,vector<int>(cols));
        queue<node>q;
        for(int r = 0; r < rows; r++){
            for(int c = 0; c < cols; c++){
                incomingEdges[r][c] = biggerNeighboursAndEdges(r,c,matrix).second;
                if(incomingEdges[r][c]==0)
                    q.push({r,c}); 
            }
        }
        int levels = 0;
        while(!q.empty()){
            int level_sz = q.size();
            while(level_sz--){
                int r = q.front().r;
                int c = q.front().c;
                q.pop();
                vector<node>biggerNeighbours = biggerNeighboursAndEdges(r,c,matrix).first;
                
                for(auto&node:biggerNeighbours){
                    if(--incomingEdges[node.r][node.c]==0)
                        q.push(node);
                }
            }
            levels++;
        }
        return levels;
    }
};
